File: apps/autopilot/src/types/session-continuity.ts
export interface SessionSummary {
  session_id: string;
  title: string;
  timestamp: number;
  phase?: string | null;
  current_step?: string | null;
  resume_hint?: string | null;
  workspace_root?: string | null;
}

export type SessionMemoryClass =
  | "ephemeral"
  | "carry_forward"
  | "pinned"
  | "governance_critical";

export type SessionCompactionMode = "manual" | "auto";

export interface SessionCompactionPolicy {
  carryPinnedOnly: boolean;
  preserveChecklistState: boolean;
  preserveBackgroundTasks: boolean;
  preserveLatestOutputExcerpt: boolean;
  preserveGovernanceBlockers: boolean;
  aggressiveTranscriptPruning: boolean;
}

export type SessionCompactionDisposition =
  | "carry_forward"
  | "retained_summary"
  | "pruned";

export type SessionCompactionResumeSafetyStatus = "protected" | "degraded";

export interface SessionCompactionResumeSafetyReceipt {
  status: SessionCompactionResumeSafetyStatus;
  reasons: string[];
}

export interface SessionCompactionMemoryItem {
  key: string;
  label: string;
  memoryClass: SessionMemoryClass;
  values: string[];
}

export interface SessionCompactionPruneDecision {
  key: string;
  label: string;
  disposition: SessionCompactionDisposition;
  detailCount: number;
  rationale: string;
  summary: string;
  examples: string[];
}

export interface SessionCompactionCarryForwardState {
  workspaceRoot?: string | null;
  pinnedFiles: string[];
  explicitIncludes: string[];
  explicitExcludes: string[];
  checklistLabels: string[];
  backgroundTaskLabels: string[];
  blockedOn?: string | null;
  pendingDecisionContext?: string | null;
  latestArtifactOutcome?: string | null;
  executionTargets: string[];
  latestOutputExcerpt?: string | null;
  memoryItems: SessionCompactionMemoryItem[];
}

export interface SessionCompactionPreview {
  sessionId: string;
  title: string;
  phase?: string | null;
  policy: SessionCompactionPolicy;
  preCompactionSpan: string;
  summary: string;
  resumeAnchor: string;
  carriedForwardState: SessionCompactionCarryForwardState;
  resumeSafety: SessionCompactionResumeSafetyReceipt;
  pruneDecisions: SessionCompactionPruneDecision[];
}

export interface SessionCompactionRecord {
  compactionId: string;
  sessionId: string;
  title: string;
  compactedAtMs: number;
  mode: SessionCompactionMode;
  phase?: string | null;
  policy: SessionCompactionPolicy;
  preCompactionSpan: string;
  summary: string;
  resumeAnchor: string;
  carriedForwardState: SessionCompactionCarryForwardState;
  resumeSafety: SessionCompactionResumeSafetyReceipt;
  pruneDecisions: SessionCompactionPruneDecision[];
}

export interface SessionCompactionRecommendation {
  shouldCompact: boolean;
  reasonLabels: string[];
  recommendedPolicy: SessionCompactionPolicy;
  recommendedPolicyLabel: string;
  recommendedPolicyReasonLabels: string[];
  resumeSafeguardLabels: string[];
  historyCount: number;
  eventCount: number;
  artifactCount: number;
  pinnedFileCount: number;
  explicitIncludeCount: number;
  idleAgeMs: number;
  blockedAgeMs?: number | null;
}

export interface SessionDurabilityPortfolio {
  retainedSessionCount: number;
  compactedSessionCount: number;
  replayReadySessionCount: number;
  uncompactedSessionCount: number;
  staleCompactionCount: number;
  degradedCompactionCount: number;
  recommendedCompactionCount: number;
  compactedWithoutTeamMemoryCount: number;
  teamMemoryEntryCount: number;
  teamMemoryCoveredSessionCount: number;
  teamMemoryRedactedSessionCount: number;
  teamMemoryReviewRequiredSessionCount: number;
  coverageSummary: string;
  teamMemorySummary: string;
  attentionSummary: string;
  attentionLabels: string[];
}

export interface SessionCompactionSnapshot {
  generatedAtMs: number;
  activeSessionId?: string | null;
  activeSessionTitle?: string | null;
  policyForActive: SessionCompactionPolicy;
  recordCount: number;
  latestForActive?: SessionCompactionRecord | null;
  previewForActive?: SessionCompactionPreview | null;
  recommendationForActive?: SessionCompactionRecommendation | null;
  durabilityPortfolio?: SessionDurabilityPortfolio;
  records: SessionCompactionRecord[];
}

export type TeamMemoryScopeKind = "workspace" | "session";

export type TeamMemorySyncStatus = "synced" | "redacted" | "review_required";

export interface TeamMemoryRedactionSummary {
  redactionCount: number;
  redactedFields: string[];
  redactionVersion: string;
}

export interface TeamMemorySyncEntry {
  entryId: string;
  sessionId: string;
  sessionTitle: string;
  syncedAtMs: number;
  scopeKind: TeamMemoryScopeKind;
  scopeId: string;
  scopeLabel: string;
  actorId: string;
  actorLabel: string;
  actorRole: string;
  syncStatus: TeamMemorySyncStatus;
  reviewSummary: string;
  omittedGovernanceItemCount: number;
  resumeAnchor: string;
  preCompactionSpan: string;
  summary: string;
  sharedMemoryItems: SessionCompactionMemoryItem[];
  redaction: TeamMemoryRedactionSummary;
}

export interface TeamMemorySyncSnapshot {
  generatedAtMs: number;
  activeSessionId?: string | null;
  activeScopeId?: string | null;
  activeScopeKind?: TeamMemoryScopeKind | null;
  activeScopeLabel?: string | null;
  entryCount: number;
  redactedEntryCount: number;
  reviewRequiredCount: number;
  summary: string;
  entries: TeamMemorySyncEntry[];
}

export interface SessionRewindCandidate {
  sessionId: string;
  title: string;
  timestamp: number;
  phase?: string | null;
  currentStep?: string | null;
  resumeHint?: string | null;
  workspaceRoot?: string | null;
  isCurrent: boolean;
  isLastStable: boolean;
  actionLabel: string;
  previewHeadline: string;
  previewDetail: string;
  discardSummary: string;
}

export interface SessionRewindSnapshot {
  activeSessionId?: string | null;
  activeSessionTitle?: string | null;
  lastStableSessionId?: string | null;
  candidates: SessionRewindCandidate[];
}

export interface SessionHookReceiptSummary {
  title: string;
  timestampMs: number;
  toolName: string;
  status: string;
  summary: string;
}

export interface SessionHookRecord {
  hookId: string;
  entryId?: string | null;
  label: string;
  ownerLabel: string;
  sourceLabel: string;
  sourceKind: string;
  sourceUri?: string | null;
  contributionPath?: string | null;
  triggerLabel: string;
  enabled: boolean;
  statusLabel: string;
  trustPosture: string;
  governedProfile: string;
  authorityTierLabel: string;
  availabilityLabel: string;
  sessionScopeLabel: string;
  whyActive: string;
}

export interface SessionHookSnapshot {
  generatedAtMs: number;
  sessionId?: string | null;
  workspaceRoot?: string | null;
  activeHookCount: number;
  disabledHookCount: number;
  runtimeReceiptCount: number;
  approvalReceiptCount: number;
  hooks: SessionHookRecord[];
  recentReceipts: SessionHookReceiptSummary[];
}

export interface SessionBranchRecord {
  branchName: string;
  upstreamBranch?: string | null;
  isCurrent: boolean;
  aheadCount: number;
  behindCount: number;
  lastCommit?: string | null;
}

export interface SessionWorktreeRecord {
  path: string;
  branchName?: string | null;
  head?: string | null;
  lastCommit?: string | null;
  changedFileCount: number;
  dirty: boolean;
  isCurrent: boolean;
  locked: boolean;
  lockReason?: string | null;
  prunable: boolean;
  pruneReason?: string | null;
  statusLabel: string;
  statusDetail: string;
}

export interface SessionBranchSnapshot {
  generatedAtMs: number;
  sessionId?: string | null;
  workspaceRoot?: string | null;
  isRepo: boolean;
  repoLabel?: string | null;
  currentBranch?: string | null;
  upstreamBranch?: string | null;
  lastCommit?: string | null;
  aheadCount: number;
  behindCount: number;
  changedFileCount: number;
  dirty: boolean;
  worktreeRiskLabel: string;
  worktreeRiskDetail: string;
  recentBranches: SessionBranchRecord[];
  worktrees: SessionWorktreeRecord[];
}

export interface SessionRemoteEnvBinding {
  key: string;
  valuePreview: string;
  sourceLabel: string;
  scopeLabel: string;
  provenanceLabel: string;
  secret: boolean;
  redacted: boolean;
}

export interface SessionRemoteEnvSnapshot {
  generatedAtMs: number;
  sessionId?: string | null;
  workspaceRoot?: string | null;
  focusedScopeLabel: string;
  governingSourceLabel: string;
  postureLabel: string;
  postureDetail: string;
  bindingCount: number;
  controlPlaneBindingCount: number;
  processBindingCount: number;
  overlappingBindingCount: number;
  secretBindingCount: number;
  redactedBindingCount: number;
  notes: string[];
  bindings: SessionRemoteEnvBinding[];
}

export interface SessionServerSessionRecord {
  sessionId: string;
  title: string;
  timestamp: number;
  sourceLabel: string;
  presenceState: string;
  presenceLabel: string;
  resumeHint?: string | null;
  workspaceRoot?: string | null;
}

export interface SessionServerSnapshot {
  generatedAtMs: number;
  sessionId?: string | null;
  workspaceRoot?: string | null;
  rpcUrl: string;
  rpcSourceLabel: string;
  continuityModeLabel: string;
  continuityStatusLabel: string;
  continuityDetail: string;
  kernelConnectionLabel: string;
  kernelConnectionDetail: string;
  explicitRpcTarget: boolean;
  remoteKernelTarget: boolean;
  kernelReachable: boolean;
  remoteHistoryAvailable: boolean;
  localSessionCount: number;
  remoteSessionCount: number;
  mergedSessionCount: number;
  remoteOnlySessionCount: number;
  overlappingSessionCount: number;
  remoteAttachableSessionCount: number;
  remoteHistoryOnlySessionCount: number;
  currentSessionVisibleRemotely: boolean;
  currentSessionContinuityState: string;
  currentSessionContinuityLabel: string;
  currentSessionContinuityDetail: string;
  notes: string[];
  recentRemoteSessions: SessionServerSessionRecord[];
}

File: apps/autopilot/src-tauri/src/models/session_continuity.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use super::session::{AgentPhase, AgentTask};
use super::session_compaction::SessionSummary;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionProjection {
    pub task: Option<AgentTask>,
    pub sessions: Vec<SessionSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionRewindCandidate {
    pub session_id: String,
    pub title: String,
    pub timestamp: u64,
    #[serde(default)]
    pub phase: Option<AgentPhase>,
    #[serde(default)]
    pub current_step: Option<String>,
    #[serde(default)]
    pub resume_hint: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
    pub is_current: bool,
    pub is_last_stable: bool,
    pub action_label: String,
    pub preview_headline: String,
    pub preview_detail: String,
    pub discard_summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionRewindSnapshot {
    #[serde(default)]
    pub active_session_id: Option<String>,
    #[serde(default)]
    pub active_session_title: Option<String>,
    #[serde(default)]
    pub last_stable_session_id: Option<String>,
    pub candidates: Vec<SessionRewindCandidate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionHookReceiptSummary {
    pub title: String,
    pub timestamp_ms: u64,
    pub tool_name: String,
    pub status: String,
    pub summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionHookRecord {
    pub hook_id: String,
    #[serde(default)]
    pub entry_id: Option<String>,
    pub label: String,
    pub owner_label: String,
    pub source_label: String,
    pub source_kind: String,
    #[serde(default)]
    pub source_uri: Option<String>,
    #[serde(default)]
    pub contribution_path: Option<String>,
    pub trigger_label: String,
    pub enabled: bool,
    pub status_label: String,
    pub trust_posture: String,
    pub governed_profile: String,
    pub authority_tier_label: String,
    pub availability_label: String,
    pub session_scope_label: String,
    pub why_active: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionHookSnapshot {
    pub generated_at_ms: u64,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
    pub active_hook_count: usize,
    pub disabled_hook_count: usize,
    pub runtime_receipt_count: usize,
    pub approval_receipt_count: usize,
    pub hooks: Vec<SessionHookRecord>,
    #[serde(default)]
    pub recent_receipts: Vec<SessionHookReceiptSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionBranchRecord {
    pub branch_name: String,
    #[serde(default)]
    pub upstream_branch: Option<String>,
    pub is_current: bool,
    #[serde(default)]
    pub ahead_count: u32,
    #[serde(default)]
    pub behind_count: u32,
    #[serde(default)]
    pub last_commit: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionWorktreeRecord {
    pub path: String,
    #[serde(default)]
    pub branch_name: Option<String>,
    #[serde(default)]
    pub head: Option<String>,
    #[serde(default)]
    pub last_commit: Option<String>,
    #[serde(default)]
    pub changed_file_count: usize,
    #[serde(default)]
    pub dirty: bool,
    #[serde(default)]
    pub is_current: bool,
    #[serde(default)]
    pub locked: bool,
    #[serde(default)]
    pub lock_reason: Option<String>,
    #[serde(default)]
    pub prunable: bool,
    #[serde(default)]
    pub prune_reason: Option<String>,
    pub status_label: String,
    pub status_detail: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionBranchSnapshot {
    pub generated_at_ms: u64,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
    pub is_repo: bool,
    #[serde(default)]
    pub repo_label: Option<String>,
    #[serde(default)]
    pub current_branch: Option<String>,
    #[serde(default)]
    pub upstream_branch: Option<String>,
    #[serde(default)]
    pub last_commit: Option<String>,
    #[serde(default)]
    pub ahead_count: u32,
    #[serde(default)]
    pub behind_count: u32,
    #[serde(default)]
    pub changed_file_count: usize,
    #[serde(default)]
    pub dirty: bool,
    pub worktree_risk_label: String,
    pub worktree_risk_detail: String,
    #[serde(default)]
    pub recent_branches: Vec<SessionBranchRecord>,
    #[serde(default)]
    pub worktrees: Vec<SessionWorktreeRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionRemoteEnvBinding {
    pub key: String,
    pub value_preview: String,
    pub source_label: String,
    pub scope_label: String,
    pub provenance_label: String,
    #[serde(default)]
    pub secret: bool,
    #[serde(default)]
    pub redacted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionRemoteEnvSnapshot {
    pub generated_at_ms: u64,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
    pub focused_scope_label: String,
    pub governing_source_label: String,
    pub posture_label: String,
    pub posture_detail: String,
    #[serde(default)]
    pub binding_count: usize,
    #[serde(default)]
    pub control_plane_binding_count: usize,
    #[serde(default)]
    pub process_binding_count: usize,
    #[serde(default)]
    pub overlapping_binding_count: usize,
    #[serde(default)]
    pub secret_binding_count: usize,
    #[serde(default)]
    pub redacted_binding_count: usize,
    #[serde(default)]
    pub notes: Vec<String>,
    #[serde(default)]
    pub bindings: Vec<SessionRemoteEnvBinding>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionServerSessionRecord {
    pub session_id: String,
    pub title: String,
    pub timestamp: u64,
    pub source_label: String,
    #[serde(default)]
    pub presence_state: String,
    #[serde(default)]
    pub presence_label: String,
    #[serde(default)]
    pub resume_hint: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct SessionServerSnapshot {
    pub generated_at_ms: u64,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub workspace_root: Option<String>,
    pub rpc_url: String,
    pub rpc_source_label: String,
    pub continuity_mode_label: String,
    pub continuity_status_label: String,
    pub continuity_detail: String,
    pub kernel_connection_label: String,
    pub kernel_connection_detail: String,
    #[serde(default)]
    pub explicit_rpc_target: bool,
    #[serde(default)]
    pub remote_kernel_target: bool,
    #[serde(default)]
    pub kernel_reachable: bool,
    #[serde(default)]
    pub remote_history_available: bool,
    #[serde(default)]
    pub local_session_count: usize,
    #[serde(default)]
    pub remote_session_count: usize,
    #[serde(default)]
    pub merged_session_count: usize,
    #[serde(default)]
    pub remote_only_session_count: usize,
    #[serde(default)]
    pub overlapping_session_count: usize,
    #[serde(default)]
    pub remote_attachable_session_count: usize,
    #[serde(default)]
    pub remote_history_only_session_count: usize,
    #[serde(default)]
    pub current_session_visible_remotely: bool,
    #[serde(default)]
    pub current_session_continuity_state: String,
    #[serde(default)]
    pub current_session_continuity_label: String,
    #[serde(default)]
    pub current_session_continuity_detail: String,
    #[serde(default)]
    pub notes: Vec<String>,
    #[serde(default)]
    pub recent_remote_sessions: Vec<SessionServerSessionRecord>,
}
